### DISCUSSION TITLE
classes/class_projectsettings

### DISCUSSION THREAD

**Discussion by giscus** [godotengine/godot-docs-user-notes#240] [User-contributed notes]
# classes/class_projectsettings

Inherits: Object Stores globally-accessible variables. Description: Stores variables that can be accessed from everywhere. Use get_setting, set_setting or has_setting to access them. Variables stor...

https://docs.godotengine.org/en/latest/classes/class_projectsettings.html

**Comment by lben** [comment]
Are display/window/size/viewport_width and display/window/size/viewport_height in pixels? I have a sprite that has 512px of height, I set it up to display/window/size/viewport_height=512 but when I start the scene it draws the window smaller than 512px, also it looks as it should in the editor (with 100% zoom) but not when being run, why? This is unintuitive behavior

**Reply by Calinou** [reply]
Yes, they are in pixels. Note that zoom in the 2D editor is scaled according to the **Editor Scale** setting, so if you're using 200% editor scale, 100% zoom in 2D will actually be 200% in screen pixels. (This is consistent with how zoom percentages are displayed in image editors and web browsers.)

Also, Godot currently doesn't scale the window scale according to the OS display scaling, so the window will appear to be smaller than usual at scales above 100%. See https://github.com/godotengine/godot-proposals/issues/7968.

**Comment by Gnumaru** [comment]
The description of load_resource_pack() notes that "DirAccess will not show changes made to the contents of res:// after calling this function". This is not true. I have just tested this in godot 4.4 and after loading a pck file with ProjectSettings.load_resource_pack, the contents of res:// where updated and I could list them with open() and then list_dir_begin() and get_next().

And one more thing about load_resource_pack(). How is it meant to be used with zip files? How should be the workflow and in which scenario would it be preferable to use zips over pcks? I've made some tests and noticed that I could zip the entire contents of a godot project, load it with load_resource_pack and it seems to work (but I have only made simple tests so far). But I need to zip the contents of the root directory of the project. That is, the files and directories at the root of the zip should be those found on "res://", if I where to zip the directory of the project instead of the contents of the project directory, in a similar manner as when you use "project > pack project as zip" (which is meant for archival purposes, not for loading as resource pack), it wouldn't work.

Edit: I can't believe I have never noticed that the export dialog button for exporting only the pck reads as "export pck **/zip** ". That is the "workflow" I mentioned above, the expected way by which one can generate zips equivalent to pcks meant to be loaded with load_resource_pack. I'm not trying to read a generic zip file for generic purposes, I'm also already using the ZIPReader class to check the contents of zip files to diferentiate between valid zip files meant to be loaded with load_resource_pack and other ones that are not equivalent to a pck file. But my doubt about "in which scenario would it be preferable to use zips over pcks" remains. Since it is not possible to, for example, swap a png inside of a zip file (in order to modding a game) because what godot games uses at runtime are actually the imported files in .godot/imported, then I can't see the benefit of exporting as zip instead of pck, unless the game itself was designed to not use the imported files, but to always load the source files with something like "var i = Image.new(); i.load("res://thing.png"); var t = ImageTexture.create_from_image(i); %Sprite.texture = t;"

**Reply by Calinou** [reply]
If you want to manually read files from a ZIP archive (without Godot's virtual filesytem handling), use the ZIPReader class.

`load_resource_pack()` does not support changing the root path so that it reads from a root folder instead (although it could be argued such behavior causes ambiguity for modders, since PCK/ZIP loading will no longer work in an universal manner across projects that load resource packs at runtime).

**Comment by Elmapul** [comment]
downloaded the lastest stable build of godot, and the function OS.localize_path("") dont seems to exist, im not sure if it does what i think it does...
but it seems to be the feature that i need right now.

**Reply by Elmapul** [reply]
never mind, its on  ProjectSettings not on OS

**Comment by mazekasandile8-droid** [comment]
I think my doc has bug, page's are too long, and characters are written in separately line

**Reply by Calinou** [reply]
Try resizing the editor window. I've seen this issue reported on GitHub already, but I can't find it anymore.

**Comment by BinaryGleam** [comment]
I think an example of syntax should be provided to the [override config file section](https://docs.godotengine.org/en/latest/classes/class_projectsettings.html#description), specifically for some interesting use case. Right now I did it for the editor but you should use the method stipulated in [here](https://docs.godotengine.org/en/stable/tutorials/export/feature_tags.html#overriding-project-settings) so it's not a correct use case. But at least it shows how to do force window mode specifically for the editor.

```
[display]
window/size/mode.editor = 0
```

If anyone has an actual concrete use case that would show how to specify settings between 2 specific feature tag, that would be awesome. Sadly I'm not one with a specific case like that

**Comment by NuclearPhoenixx** [comment]
Why is there a `get_setting_with_override`, but not a `set_setting_with_override`?

**Reply by Calinou** [reply]
`get_setting_with_override()` is meant to say that you want to get a project setting's value, but take any feature tag overrides into account. `get_setting()` will get you the setting directly, disregarding feature tags. (That said, you can still query a specific feature tag's value using a `.feature` suffix to the project setting name, e.g. `.macos`.)

In contrast, there is no need to do that with `set_setting()`. You can use the feature tag name as a suffix of the project setting directly: `set_setting("section/key.macos", "value")`

**Reply by NuclearPhoenixx** [reply]
Yeah of course you can hardcode the feature tags in both get or set, but what if I want to get a setting using `get_setting_with_override` and then depending on its value change the same tag to a different value using `set_setting_with_override`. Currently, I'd have to get the setting with override, then detect the correct feature tag and hardcode that into `set_setting`.

I'm using it to change settings for the rendering driver in the options menu. If I want to change the driver on Windows for example between Vulkan an D3D12, I have to first detect the current OS, check if it's Windows to allow D3D12 in the first place, and then manually append `.windows` to the `set_setting` setting key string. Instead, it would be much easier to just use `set_setting_with_override` and get rid of the redundant code. I hope I explained that clear enough.

**Reply by Calinou** [reply]
The issue is that each platform matches multiple feature tags, so the engine wouldn't be able to know which flag you want to specify:

- Platform
- Platform type (desktop/mobile/web)
- CPU architecture
- Debug/release
- Editor/template
- ...

It is possible to override a single setting with different feature tag types, although priority is not specified (and it's not possible to nest feature tags, e.g. `windows.debug`).

**Reply by NuclearPhoenixx** [reply]
Ok I see, very good point! Though, the same would be true about the get function, wouldn't it? A potential set func would of course need to have the same logic to target feature tags as the getter.

**Reply by abrusle** [reply]
What order/priority logic does the engine use when multiple feature tags are applicable? (e.g. `*.debug` vs. `*.64` vs. `*.single`)

**Reply by Calinou** [reply]
The priority is not defined as far as I know, so this is considered undefined behavior. It needs further testing, but I wouldn't rely on any specific priority until it's documented.

**Comment by v1993** [comment]
Does enabling `physics/Xd/run_on_separate_thread` make `_physics_process` methods run on a separate thread as well? What about `_integrate_forces`? Doesn't seem so, but better be safe than sorry.

**Comment by TerrellAW** [comment]
How can I use .cfg file to enable a plugin? It does not work.

**Reply by Calinou** [reply]
Please don't ask support questions here, as user notes are not designed for this. Use one of the other [community channels](https://godotengine.org/community) instead.
